### PR TITLE
`GraphSAGE` example with `LinkNeighborLoader`

### PR DESCRIPTION
**Pull request by scigeek72**
Replaced NeighborSampler by LinkNeighborLoader and made the necessary changes

### CODE REVIEW

File: examples/graph_sage_unsup.py [hunk 1]
```diff
@@ -78,17 +63,16 @@ def train():
...
+        h = model(loader)
+        src_nodes = loader.edge_label_index[0]
+        dst_nodes = loader.edge_label_index[1]
+        out = h[src_nodes] * h[dst_nodes]
+        loss = F.binary_cross_entropy_with_logits(out.sum(dim=-1),
+                                                  loader.edge_label)
 
-        pos_loss = F.logsigmoid((out * pos_out).sum(-1)).mean()
-        neg_loss = F.logsigmoid(-(out * neg_out).sum(-1)).mean()
-        loss = -pos_loss - neg_loss
```
Comment (wsad1) [inline examples/graph_sage_unsup.py:91]: is the new loss equivalent to the previous loss.

File: examples/graph_sage_unsup.py [hunk 2]
```diff
@@ -49,10 +33,11 @@ def __init__(self, in_channels, hidden_channels, num_layers):
...
 
-    def forward(self, x, adjs):
-        for i, (edge_index, _, size) in enumerate(adjs):
-            x_target = x[:size[1]]  # Target nodes are always placed first.
-            x = self.convs[i]((x, x_target), edge_index)
+    def forward(self, data):
+
+        x, edge_index = data.x, data.edge_index
+        for i in range(self.num_layers):
+            x = self.convs[i]((x, x), edge_index)
```
Comment (wsad1) [inline examples/graph_sage_unsup.py:40]:
```suggestion
            x = self.convs[i](x, edge_index)
```
I guess this should also work.

**Review by wsad1 — COMMENTED** [review]
Thanks for updating the example. Left a few comments. Will take a look again later.

File: examples/graph_sage_unsup.py [hunk 3]
```diff
@@ -4,40 +4,24 @@
...
 from sklearn.linear_model import LogisticRegression
-from torch_cluster import random_walk
 
 import torch_geometric.transforms as T
 from torch_geometric.datasets import Planetoid
-from torch_geometric.loader import NeighborSampler as RawNeighborSampler
+from torch_geometric.loader import LinkNeighborLoader
 from torch_geometric.nn import SAGEConv
 
+#from torch_cluster import random_walk
```
Comment (rusty1s) [inline examples/graph_sage_unsup.py:13]: Remove.

File: examples/graph_sage_unsup.py [hunk 4]
```diff
@@ -4,40 +4,24 @@
...
-
-        batch = torch.cat([batch, pos_batch, neg_batch], dim=0)
-        return super().sample(batch)
-
-
-train_loader = NeighborSampler(data.edge_index, sizes=[10, 10], batch_size=256,
-                               shuffle=True, num_nodes=data.num_nodes)
+train_loader = LinkNeighborLoader(data, batch_size=256, shuffle=True,
+                                  neg_sampling_ratio=1.0,
+                                  num_neighbors=[10, 10])
```
Comment (rusty1s) [inline examples/graph_sage_unsup.py:24]: Can we also increase `num_workers` to `6` (or similar)?

File: examples/graph_sage_unsup.py [hunk 5]
```diff
@@ -49,10 +33,11 @@ def __init__(self, in_channels, hidden_channels, num_layers):
             in_channels = in_channels if i == 0 else hidden_channels
             self.convs.append(SAGEConv(in_channels, hidden_channels))
 
-    def forward(self, x, adjs):
-        for i, (edge_index, _, size) in enumerate(adjs):
-            x_target = x[:size[1]]  # Target nodes are always placed first.
-            x = self.convs[i]((x, x_target), edge_index)
+    def forward(self, data):
+
+        x, edge_index = data.x, data.edge_index
```
Comment (rusty1s) [inline examples/graph_sage_unsup.py:38]:
```suggestion
    def forward(self, x, edge_index):
```

File: examples/graph_sage_unsup.py (hunk 1 above)
Comment (rusty1s) [reply examples/graph_sage_unsup.py:91]: Yes :)

**Review by rusty1s — APPROVED** [review]
I like it. Thanks!

File: examples/graph_sage_unsup.py [hunk 6]
```diff
@@ -108,11 +92,11 @@ def test():
     val_acc = clf.score(out[data.val_mask], data.y[data.val_mask])
     test_acc = clf.score(out[data.test_mask], data.y[data.test_mask])
 
-    return val_acc, test_acc
+    return val_acc, test_acc, out
```
[the remark this answers is not in the document]
Comment (scigeek72) [reply examples/graph_sage_unsup.py:95]: `val_acc` and `test_acc` are reduced when I ran it. Also it was little bit slower than the original examples (but that could be because of my machine).

File: examples/graph_sage_unsup.py (hunk 4 above)
Comment (scigeek72) [reply examples/graph_sage_unsup.py:24]: I tried to use `num_workers = 6` but was getting an error ( on a MacBook Pro, no cuda support). So left it unchanged.

File: examples/graph_sage_unsup.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (wsad1) [reply examples/graph_sage_unsup.py:93]: Thanks.

File: examples/graph_sage_unsup.py [hunk 7]
```diff
@@ -108,11 +90,11 @@ def test():
...
```
Comment (wsad1) [inline examples/graph_sage_unsup.py:93]:
```suggestion
    return val_acc, test_acc
```

**Review by wsad1 — APPROVED** [review]
